pkg/policy/directory/watcher.go

```diff
@@ -0,0 +1,217 @@
+		labelsArr = append(labelsArr, srcLabel)
+	}
+
+	return labelsArr
+}
+
+// read cilium network policy yaml file and convert to policy object and
+// add rules to policy engine.
+func (p *policyWatcher) addToPolicyEngine(cnpFilePath string) error {
+	// read from file and convert to cnp object
+	cnp, err := p.translateToCNPObject(cnpFilePath)
+	if err != nil {
+		return err
+	}
+
+	cnp.ObjectMeta.Name = filepath.Base(cnpFilePath)
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindFile,
+		cnp.ObjectMeta.Namespace,
+		cnp.ObjectMeta.Name,
+	)
+
+	// convert to rules
+	rules, err := cnp.Parse()
+	if err != nil {
+		return err
+	}
+
+	//update labels
+	lbls := getLabels(cnp.ObjectMeta.Name, cnp)
+	for _, r := range rules {
+		r.Labels = lbls
+	}
+
+	// add to policy engine
+	_, err = p.policyManager.PolicyAdd(rules, &policy.AddOptions{
+		ReplaceByResource: true,
+		Source:            source.Directory,
+		Resource:          resourceID,
+	})
+
+	if err == nil {
+		p.fileNameToCnpCache[cnp.ObjectMeta.Name] = cnp
+	}
+
+	return err
+}
+
+func (p *policyWatcher) deleteFromPolicyEngine(cnpFilePath string) error {
+	fileName := filepath.Base(cnpFilePath)
+	cnp := p.fileNameToCnpCache[fileName]
+	if cnp == nil {
+		p.log.WithField("file", fileName).Warn("map entry doesn't exist")
+		return nil
+	}
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindFile,
+		cnp.ObjectMeta.Namespace,
+		fileName,
+	)
+	_, err := p.policyManager.PolicyDelete(getLabels(fileName, cnp), &policy.DeleteOptions{
+		Source:   source.Directory,
+		Resource: resourceID})
+
+	delete(p.fileNameToCnpCache, fileName)
+	return err
+}
+
+func (p *policyWatcher) watchDirectory(ctx context.Context) {
+	go func() {
+		p.log.Info("Directory policy watcher started")
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			p.log.WithField("Err", err).Error("Initializing NewWatcher failed")
+			return
+		}
+		defer watcher.Close()
+
+		dir := p.config.StaticCiliumNetworkPolicyPath
+		err = watcher.Add(dir)
+		if err != nil {
+			p.log.WithFields(logrus.Fields{"Dir": dir, "err": err}).Error("Directory Watcher Add failed")
+			return
+		}
+
+		files, err := os.ReadDir(dir)
+		if err != nil {
+			p.log.WithFields(logrus.Fields{"Directory": dir, "err": err}).Warn("Error reading CNP directory")
+			return
+		}
+
+		// Read existing cnp files before watcher was added
+		for _, f := range files {
+			err := p.addToPolicyEngine(filepath.Join(dir, f.Name()))
+			if err != nil {
+				p.log.WithFields(logrus.Fields{"file": f.Name(), "err": err}).Error("Error while adding network policy to policy engine")
+			} else {
+
+			}
+			reportCNPChangeMetrics(err)
+		}
+		close(p.readStatus)
+		// Listen for file add, update, rename and delete
+		for {
+			select {
+			case event := <-watcher.Events:
+				if !utf8.ValidString(event.Name) {
```

Comment:
Given that we now use the filename as the CNP name, I wonder if we want to enforce that the filename is a [valid CNP name](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names). We can always relax later, but that way we can be sure there is no breaking change if somehow `cnp.Parse()` becomes more restrictive in the future.

To validate if a filename is a valid CNP name, you can use `validation.IsDNS1123Subdomain(event.Name)`

Reply:
Now that we validate with `validation.IsDNS1123Subdomain` this UTF8 check is no longer needed. I'd remove it.